// html-validate/build/rules/element-case.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const context_1 = require("../context");
const rule_1 = require("../rule");
const defaults = {
    style: "lowercase",
};
class ElementCase extends rule_1.Rule {
    constructor(options) {
        super(Object.assign({}, defaults, options));
        [this.pattern, this.lettercase] = parseStyle(this.options.style);
    }
    documentation() {
        return {
            description: `Element tagname must be ${this.options.style}.`,
            url: rule_1.ruleDocumentationUrl(__filename),
        };
    }
    setup() {
        this.on("tag:open", (event) => {
            const letters = event.target.tagName.replace(/[^a-z]+/gi, "");
            if (!letters.match(this.pattern)) {
                const location = context_1.sliceLocation(event.location, 1);
                this.report(event.target, `Element "${event.target.tagName}" should be ${this.lettercase}`, location);
            }
        });
    }
}
function parseStyle(style) {
    switch (style.toLowerCase()) {
        case "lowercase":
            return [/^[a-z]*$/, "lowercase"];
        case "uppercase":
            return [/^[A-Z]*$/, "uppercase"];
        default:
            throw new Error(`Invalid style "${style}" for "element-case" rule`);
    }
}
module.exports = ElementCase;
