// file-it/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var file_1 = require("./managers/file");
var universalify = require("universalify");
var fileIt = {
    makeDirSync: file_1._makeDirSync,
    appendJsonFileSync: file_1._appendJsonFileSync,
    readContentFileAsync: universalify.fromPromise(file_1._readContentFileAsync),
    readContentFileSync: file_1._readContentFileSync,
    readJsonFile: universalify.fromPromise(file_1._readJsonFileAsync),
    readJsonFileSync: file_1._readJsonFileSync,
    writeContentFile: universalify.fromPromise(file_1._writeContentFileAsync),
    writeContentFileSync: file_1._writeContentFileSync,
    writeJsonFile: universalify.fromPromise(file_1._writeJsonFileAsync),
    writeJsonFileSync: file_1._writeJsonFileSync,
    setJsonValue: file_1._setJsonValue,
    getJsonValue: file_1._getJsonValue,
    readJsonArraySync: file_1._readJsonArraySync,
    readJsonLinesSync: file_1._readJsonLinesSync,
    findSortedJsonElement: file_1._findSortedJsonElement
};
module.exports = fileIt;
