// html-validate/build/reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("./config");
class Reporter {
    constructor() {
        this.result = {};
    }
    /**
     * Merge two or more reports into a single one.
     */
    static merge(reports) {
        const valid = reports.every(report => report.valid);
        const merged = {};
        reports.forEach((report) => {
            report.results.forEach((result) => {
                const key = result.filePath;
                if (key in merged) {
                    merged[key].messages = [].concat(merged[key].messages, result.messages);
                }
                else {
                    merged[key] = Object.assign({}, result);
                }
            });
        });
        return { valid, results: Object.keys(merged).map(key => merged[key]) };
    }
    add(rule, message, severity, location, context) {
        if (!(location.filename in this.result)) {
            this.result[location.filename] = [];
        }
        this.result[location.filename].push({
            ruleId: rule.name,
            severity,
            message,
            offset: location.offset,
            line: location.line,
            column: location.column,
            size: location.size || 0,
            context,
        });
    }
    addManual(filename, message) {
        if (!(filename in this.result)) {
            this.result[filename] = [];
        }
        this.result[filename].push(message);
    }
    save(sources) {
        return {
            valid: this.isValid(),
            results: Object.keys(this.result).map(filePath => {
                const messages = [].concat(this.result[filePath]).sort(messageSort);
                const source = (sources || []).find((source) => source.filename === filePath);
                return {
                    filePath,
                    messages,
                    errorCount: countErrors(messages),
                    warningCount: countWarnings(messages),
                    source: source ? source.originalData || source.data : null,
                };
            }),
        };
    }
    isValid() {
        const numErrors = Object.values(this.result).reduce((sum, messages) => {
            return sum + countErrors(messages);
        }, 0);
        return numErrors === 0;
    }
}
exports.Reporter = Reporter;
function countErrors(messages) {
    return messages.filter(m => m.severity === config_1.Severity.ERROR).length;
}
function countWarnings(messages) {
    return messages.filter(m => m.severity === config_1.Severity.WARN).length;
}
function messageSort(a, b) {
    if (a.line < b.line) {
        return -1;
    }
    if (a.line > b.line) {
        return 1;
    }
    if (a.column < b.column) {
        return -1;
    }
    if (a.column > b.column) {
        return 1;
    }
    return 0;
}
exports.default = Reporter;
