// git-urls/dist/src/host/basicHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BasicHost {
    constructor() {
        this.regex = /(https?:\/\/|git@)([^\/:]+)(?:\/|:)([^\/:]+)(?:\/|:)([^\/:]+)/;
    }
    parse(info) {
        const matches = this.regex.exec(info.remoteUrl);
        if (!matches) {
            throw new Error(`Can't parse ${info.remoteUrl} with Default rules`);
        }
        const schema = matches[1];
        let isHttp = false;
        if (schema.startsWith("http://")) {
            isHttp = true;
        }
        let repoName = matches[4];
        if (repoName.endsWith(".git")) {
            repoName = repoName.substring(0, repoName.lastIndexOf(".git"));
        }
        let hostName = matches[2];
        let index = hostName.indexOf('@');
        if (index !== -1) {
            hostName = hostName.substring(index + 1);
        }
        return {
            hostName: hostName,
            repoName: repoName,
            branchName: info.branchName,
            userName: matches[3],
            metadata: { "isHttp": isHttp },
        };
    }
    assemble(info) {
        return this.assembleLink(info);
    }
    assembleLink(info) {
        let prefix = "https://";
        if (info.metadata && info.metadata["isHttp"]) {
            prefix = "http://";
        }
        return `${prefix}${info.hostName}/${info.userName}/${info.repoName}/${this.separateFolder}/${info.branchName}/${info.relativefilePath}`;
    }
}
exports.default = BasicHost;
