// git-urls/dist/src/host/vsts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Vsts {
    static match(url) {
        return Vsts.urlRegex.test(url);
    }
    parse(info) {
        return {
            repoName: info.remoteUrl,
            branchName: info.branchName,
            userName: ''
        };
    }
    assemble(info) {
        const baseUrl = info.repoName.replace(Vsts.urlRegex, "https://$1.visualstudio.com/$2/_git/$3");
        const path = encodeURIComponent(`/${info.relativefilePath}`);
        let url = `${baseUrl}?path=${path}&version=GB${info.branchName}&_a=contents`;
        if (info.section && info.section.startLine && info.section.endLine) {
            url += `&lineStyle=plain&line=${info.section.startLine}&lineEnd=${info.section.endLine}`;
            if (info.section.startColumn && info.section.endColumn) {
                url += `&lineStartColumn=${info.section.startColumn}&lineEndColumn=${info.section.endColumn}`;
            }
            else {
                url += "&lineStartColumn=1&lineEndColumn=1";
            }
        }
        return url;
    }
}
/**
 * The regular expression to match the VSTS Git URL.
 * @example https://my-tenant.visualstudio.com/DefaultCollection/MyCollection/_git/my-repo
 * @example ssh://my-tenant@my-tenant.visualstudio.com:22/DefaultCollection/MyCollection/_ssh/my-repo
 */
Vsts.urlRegex = /(?:https:\/\/|ssh:\/\/)([\w-]+)@?.*\.visualstudio\.com(?:\:\d+)?\/(.+)\/(?:_git|_ssh)\/([^/]+)/i;
exports.default = Vsts;
