// html-validate/build/lexer/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TokenType;
(function (TokenType) {
    TokenType[TokenType["WHITESPACE"] = 1] = "WHITESPACE";
    TokenType[TokenType["NEWLINE"] = 2] = "NEWLINE";
    TokenType[TokenType["DOCTYPE_OPEN"] = 3] = "DOCTYPE_OPEN";
    TokenType[TokenType["DOCTYPE_VALUE"] = 4] = "DOCTYPE_VALUE";
    TokenType[TokenType["DOCTYPE_CLOSE"] = 5] = "DOCTYPE_CLOSE";
    TokenType[TokenType["TAG_OPEN"] = 6] = "TAG_OPEN";
    TokenType[TokenType["TAG_CLOSE"] = 7] = "TAG_CLOSE";
    TokenType[TokenType["ATTR_NAME"] = 8] = "ATTR_NAME";
    TokenType[TokenType["ATTR_VALUE"] = 9] = "ATTR_VALUE";
    TokenType[TokenType["TEXT"] = 10] = "TEXT";
    TokenType[TokenType["SCRIPT"] = 11] = "SCRIPT";
    TokenType[TokenType["COMMENT"] = 12] = "COMMENT";
    TokenType[TokenType["CONDITIONAL"] = 13] = "CONDITIONAL";
    TokenType[TokenType["DIRECTIVE"] = 14] = "DIRECTIVE";
    TokenType[TokenType["EOF"] = 15] = "EOF";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
