// html-validate/build/rules/wcag/h32.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const rule_1 = require("../../rule");
class H32 extends rule_1.Rule {
    documentation() {
        return {
            description: "WCAG 2.1 requires each `<form>` element to have at least one submit button.",
            url: rule_1.ruleDocumentationUrl(__filename),
        };
    }
    constructor(options) {
        super(options);
        this.name = "WCAG/H32";
    }
    setup() {
        this.on("dom:ready", (event) => {
            const forms = event.document.getElementsByTagName("form");
            forms.forEach((node) => {
                /* find submit buttons */
                for (const button of node.querySelectorAll("button,input")) {
                    const type = button.getAttribute("type");
                    if (type && type.valueMatches("submit")) {
                        return;
                    }
                }
                this.report(node, "<form> element must have a submit button");
            });
        });
    }
}
module.exports = H32;
