// vsls/vsls-contactprotocol.js
"use strict";
//----------------------
// <auto-generated>
//     Generated using the NSwag toolchain v9.10.19.0 (Newtonsoft.Json v9.0.0.0) (http://NJsonSchema.org)
// </auto-generated>
//----------------------
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var DataFormat;
(function (DataFormat) {
    DataFormat[DataFormat["Default"] = 'default'] = "Default";
    DataFormat[DataFormat["Email"] = 'email'] = "Email";
    DataFormat[DataFormat["Token"] = 'token'] = "Token";
    DataFormat[DataFormat["Path"] = 'path'] = "Path";
    DataFormat[DataFormat["Text"] = 'text'] = "Text";
    DataFormat[DataFormat["Uri"] = 'uri'] = "Uri";
    DataFormat[DataFormat["Omit"] = 'omit'] = "Omit";
    DataFormat[DataFormat["OmitDefault"] = 'omitDefault'] = "OmitDefault";
    DataFormat[DataFormat["Hide"] = 'hide'] = "Hide";
    DataFormat[DataFormat["SingleMember"] = 'singleMember'] = "SingleMember";
})(DataFormat = exports.DataFormat || (exports.DataFormat = {}));
/** Creates a decorator that saves the property format in a map attached to the class constructor. */
function dataFormat(...format) {
    return function (target, propertyKey) {
        if (!target.constructor.dataFormat) {
            target.constructor.dataFormat = {};
        }
        target.constructor.dataFormat[propertyKey] = format;
    };
}
exports.dataFormat = dataFormat;
class EventArgs {
}
exports.EventArgs = EventArgs;
/** Notification class for contact service */
class NotifyContactServiceEventArgs extends EventArgs {
}
exports.NotifyContactServiceEventArgs = NotifyContactServiceEventArgs;
/** The presence status default options */
var PresenceStatus;
(function (PresenceStatus) {
    PresenceStatus[PresenceStatus["Unknown"] = 'unknown'] = "Unknown";
    PresenceStatus[PresenceStatus["Available"] = 'available'] = "Available";
    PresenceStatus[PresenceStatus["Busy"] = 'busy'] = "Busy";
    PresenceStatus[PresenceStatus["DoNotDisturb"] = 'doNotDisturb'] = "DoNotDisturb";
    PresenceStatus[PresenceStatus["Away"] = 'away'] = "Away";
    PresenceStatus[PresenceStatus["Invisible"] = 'invisible'] = "Invisible";
    PresenceStatus[PresenceStatus["Offline"] = 'offline'] = "Offline";
})(PresenceStatus = exports.PresenceStatus || (exports.PresenceStatus = {}));
/** Contact entity */
class Contact {
}
exports.Contact = Contact;
/** Empty parameters */
class EmptyRequest {
}
exports.EmptyRequest = EmptyRequest;
/** Empty result */
class EmptyResponse {
}
exports.EmptyResponse = EmptyResponse;
/** Parameters passed during Initialize method */
class InitializeRequest {
}
exports.InitializeRequest = InitializeRequest;
/** Definition of a workspace folder */
class WorkspaceFolder {
}
exports.WorkspaceFolder = WorkspaceFolder;
/** A workspace Uri from vscode */
class WorkspaceUri {
}
exports.WorkspaceUri = WorkspaceUri;
/** Capabilities that a service would offer */
class Capabilities {
}
exports.Capabilities = Capabilities;
/** Returned on the Initialized method */
class InitializeResponse {
}
__decorate([
    dataFormat(Capabilities)
], InitializeResponse.prototype, "capabilities", void 0);
exports.InitializeResponse = InitializeResponse;
/** A presence status changed for a contact entity */
class ContactPresenceUpdate {
}
exports.ContactPresenceUpdate = ContactPresenceUpdate;
/** Batch of presence changed for multiple contacts */
class PresenceChangedNotification {
    constructor() {
        this.changes = [];
    }
}
__decorate([
    dataFormat(ContactPresenceUpdate)
], PresenceChangedNotification.prototype, "changes", void 0);
exports.PresenceChangedNotification = PresenceChangedNotification;
class InviteRequest {
}
exports.InviteRequest = InviteRequest;
class InviteReceivedNotification {
}
exports.InviteReceivedNotification = InviteReceivedNotification;
class PublishPresenceRequest {
}
exports.PublishPresenceRequest = PublishPresenceRequest;
class SelfContactNotification {
    constructor() {
        this.contact = new Contact();
    }
}
__decorate([
    dataFormat(Contact)
], SelfContactNotification.prototype, "contact", void 0);
exports.SelfContactNotification = SelfContactNotification;
class ContactsNotification {
    constructor() {
        this.contacts = [];
    }
}
__decorate([
    dataFormat(Contact)
], ContactsNotification.prototype, "contacts", void 0);
exports.ContactsNotification = ContactsNotification;
class ContactChangedNotification {
    constructor() {
        this.contact = new Contact();
    }
}
__decorate([
    dataFormat(Contact)
], ContactChangedNotification.prototype, "contact", void 0);
exports.ContactChangedNotification = ContactChangedNotification;
class ContactPresenceRequest {
    constructor() {
        this.contacts = [];
    }
}
__decorate([
    dataFormat(Contact)
], ContactPresenceRequest.prototype, "contacts", void 0);
exports.ContactPresenceRequest = ContactPresenceRequest;
class ContactResponse {
    constructor() {
        this.contact = new Contact();
    }
}
__decorate([
    dataFormat(Contact)
], ContactResponse.prototype, "contact", void 0);
exports.ContactResponse = ContactResponse;
class ContactPresenceResponse {
    constructor() {
        this.contacts = [];
    }
}
__decorate([
    dataFormat(ContactResponse)
], ContactPresenceResponse.prototype, "contacts", void 0);
exports.ContactPresenceResponse = ContactPresenceResponse;
class ContactPresenceDisposeRequest {
    constructor() {
        this.contactIds = [];
    }
}
exports.ContactPresenceDisposeRequest = ContactPresenceDisposeRequest;
var Methods;
(function (Methods) {
    Methods["RequestInitializeName"] = "initialize";
    Methods["RequestDisposeName"] = "dispose";
    Methods["RequestInviteName"] = "inviteLink";
    Methods["RequestPublishPresenceName"] = "publishPresence";
    Methods["RequestContactPresenceName"] = "contactPresenceRequest";
    Methods["RequestContactPresenceDisposeName"] = "contactPresenceDispose";
    Methods["NotifyPresenceChangedName"] = "presenceChanged";
    Methods["NotifySelfContactName"] = "selfContact";
    Methods["NotifyAvailableUsersName"] = "availableUsers";
    Methods["NotifyAvailableUsersAddedName"] = "availableUsersAdded";
    Methods["NotifyAvailableUsersRemovedName"] = "availableUsersRemoved";
    Methods["NotifyInviteReceivedName"] = "inviteLinkReceived";
    Methods["NotifyContactChangedName"] = "contactChanged";
    Methods["NotifySuggestedUsersName"] = "suggestedUsers";
    Methods["NotifyCapabilitiesUpdateName"] = "capabilitiesUpdate";
})(Methods = exports.Methods || (exports.Methods = {}));
/* tslint:disable:variable-name */
exports.ContactServiceProvider = {
    methods: ['request'],
    events: ['notified'],
};
